### PR TITLE
Added better error handling

### PR DESCRIPTION
**Pull request by K0stad1n**
Started replacing asserts with if statements to check for errors

### CODE REVIEW

**Comment by CRKatri** [comment]
Thanks for getting started on this.

File: ldid.cpp [hunk 1]
```diff
@@ -1787,13 +1792,23 @@ class Stuff {
...
 
-        _assert(PKCS12_parse(value_, password.c_str(), &key_, &cert_, &ca_) != 0);
+	if(PKCS12_parse(value_, password.c_str(), &key_, &cert_, &ca_) <= 0){
+		printf("An Error occured while parsing: \n %s\n", ERR_error_string(ERR_get_error(), NULL));
+        exit(1);
+	}
+    if(key_ == NULL || cert_ == NULL){
+        printf("An Error occured while parsing: \n %s\n You're p12 cert might not be valid", ERR_error_string(ERR_get_error(), NULL));
+        exit(1);
+    }
```
Comment (CRKatri) [inline ldid.cpp:1802]: The spacing on these lines are incorrect

File: ldid.cpp [hunk 2]
```diff
@@ -1787,13 +1792,23 @@ class Stuff {
             password = passbuf;
         }
 
-        _assert(PKCS12_parse(value_, password.c_str(), &key_, &cert_, &ca_) != 0);
+	if(PKCS12_parse(value_, password.c_str(), &key_, &cert_, &ca_) <= 0){
+		printf("An Error occured while parsing: \n %s\n", ERR_error_string(ERR_get_error(), NULL));
+        exit(1);
+	}
+    if(key_ == NULL || cert_ == NULL){
+        printf("An Error occured while parsing: \n %s\n You're p12 cert might not be valid", ERR_error_string(ERR_get_error(), NULL));
```
Comment (CRKatri) [inline ldid.cpp:1800]:
```suggestion
        printf("An Error occured while parsing: \n %s\n Your p12 cert might not be valid", ERR_error_string(ERR_get_error(), NULL));
```

**Review by CRKatri — CHANGES_REQUESTED** [review]

**Comment by K0stad1n** [comment]
@CRKatri Should there be a version bump for Ldid once error checking is done?

File: ldid.cpp [hunk 3]
```diff
@@ -1801,20 +1804,18 @@ class Stuff {
...
             exit(1);
         }
-        _assert(key_ != NULL);
-        _assert(cert_ != NULL);
 
         if (ca_ == NULL)
             ca_ = sk_X509_new_null();
         if(ca_ == NULL){
-            printf("An error occured while parsing: %s\n", ERR_error_string(ERR_get_error(), NULL));
+            printf("An error occured while parsing: \n %s\n", ERR_error_string(ERR_get_error(), NULL));
```
Comment (CRKatri) [inline ldid.cpp:1818]: Please don't re-add those newlines, I had removed them on purpose.

**Comment by CRKatri** [comment]
> Should there be a version bump for Ldid once error checking is done?

That is the current plan.

**Comment by demhademha** [comment]
We should ideally use exit(EXIT_FAILURE);

**Comment by demhademha** [comment]
And EXIT_SUCCESS

**Comment by K0stad1n** [comment]
> > Should there be a version bump for Ldid once error checking is done?
> 
> 
> 
> That is the current plan.

What should the new version number be do you think?

**Comment by CRKatri** [comment]
> > > Should there be a version bump for Ldid once error checking is done?
> > 
> > 
> > That is the current plan.
> 
> What should the new version number be do you think?

It will be `2.1.5-procursus3`. The first number is the upstream version number from git.saurik.com/ldid.git. Then we do `-procursus*` to show the Procursus patch version.

File: ldid.cpp [hunk 4]
```diff
@@ -3353,7 +3396,10 @@ int main(int argc, char *argv[]) {
         _syscall(stat(path.c_str(), &info));
 
         if (S_ISDIR(info.st_mode)) {
-            _assert(flag_S);
+            if (!flag_S) {
+                fprintf(stderr, "ldid: Only -S can be used on directores\n");
```
Comment (absidue) [inline ldid.cpp:3400]:
```suggestion
                fprintf(stderr, "ldid: Only -S can be used on directories\n");
```

**Review by absidue — CHANGES_REQUESTED** [review]

**Review by absidue — APPROVED** [review]

**Review by 1Conan — APPROVED** [review]

**Pull request merged**